### PR TITLE
Reset data coordinator after each LTFB round

### LINKED ISSUE #1597: Input layer doesn't properly switch between data readers during LTFB
While debugging #1596, I encountered errors with the message:

```
Number of samples: 244 does not match the number of responses: 200
```

The error comes from:

https://github.com/LLNL/lbann/blob/95474f70114ccef94ca6194febf9e566bfaecf43/include/lbann/io/data_buffers/generic_io_buffer.hpp#L62

I suspect the error happens because LTFB switches between execution modes without reseting background fetching in the IO buffers. If I understand correctly, the input layer works like:

```c++
void fp_compute() {
  ++m_buffer_idx;
  m_io_buffers[m_buffer_idx % m_io_buffers.size()].get_data();
  m_io_buffers[(m_buffer_idx+1) % m_io_buffers.size()].fetch_data_in_background();
}
```

If we don't reset the IO buffers when switching execution modes, then the input layer receives data that was fetched asynchronously from the wrong data reader. If an LTFB round happens at step 10 and there are 11 steps/epoch, the 11th training step will expect a partial mini-batch from the training set but receive a full mini-batch from the validation set. This mismatch triggers our error.

Do we already have machinery to reset the IO buffers? During the LTFB round, it seems that the evaluation epochs are correctly using the validation set. It's only when we finish the evaluation and go back to the middle of the training set that we appear to load the wrong data.

I'm not sure if this will help #1596. I wouldn't expect us to get divergence from a few mini-batches with incorrect data, unless it was filled with garbage.

### PR DESCRIPTION
**Pull request by timmoon10**
As discussed in #1597, the data coordinator doesn't properly switch from the validation set to the training set after an LTFB round finishes. I've adjusted the LTFB callback to match how we reset the execution mode in `sgd_training_algorithm`, e.g.:

https://github.com/LLNL/lbann/blob/9fcc5c315a079cf8a49781c22a859d915ea3d5fc/src/training_algorithms/sgd_training_algorithm.cpp#L109

This fixes #1597, but does not affect #1596.

### CODE REVIEW

File: src/callbacks/ltfb.cpp [hunk 1]
```diff
@@ -339,7 +339,8 @@ EvalType evaluate(model& m, const std::string& metric_name) {
   m.make_data_store_preloaded(execution_mode::validation);
 
   // Clean up and return metric value
-  c.set_execution_mode(original_mode);
+  m.reset_mode(c, original_mode);
+  c.get_trainer().get_data_coordinator().reset_mode(c);
```
Comment (bvanessen) [inline src/callbacks/ltfb.cpp:343]: This should be correct and we do not want to change the execution mode of the execution context as was being done in the previous code.
